simd/src/Kokkos_SIMD_NEON.hpp

```diff
@@ -0,0 +1,1108 @@
+        : m_mask(mask_arg), m_lane(lane_arg) {}
+    KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION reference
+    operator=(bool value) const {
+      switch (m_lane) {
+        case 0:
+          m_mask = vset_lane_u32(value ? 0xFFFFFFFFU : 0, m_mask, 0);
+          break;
+        case 1:
+          m_mask = vset_lane_u32(value ? 0xFFFFFFFFU : 0, m_mask, 1);
+          break;
+      }
+      return *this;
+    }
+    KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION operator bool() const {
+      switch (m_lane) {
+        case 0: return vget_lane_u32(m_mask, 0) != 0;
+        case 1: return vget_lane_u32(m_mask, 1) != 0;
+      }
+      return false;
+    }
+  };
+  using value_type = bool;
+  using abi_type   = simd_abi::neon_fixed_size<2>;
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd_mask() = default;
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION explicit simd_mask(value_type value)
+      : m_value(vdup_n_u32(value ? 0xFFFFFFFFU : 0)) {}
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION static constexpr std::size_t size() {
+    return 2;
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION constexpr explicit simd_mask(
+      uint32x2_t const& value_in)
+      : m_value(value_in) {}
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd_mask(
+      simd_mask<double, simd_abi::neon_fixed_size<2>> const& other)
+      : m_value(vqmovn_u64(static_cast<uint64x2_t>(other))) {}
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION constexpr explicit operator uint32x2_t()
+      const {
+    return m_value;
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION reference operator[](std::size_t i) {
+    return reference(m_value, int(i));
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION value_type
+  operator[](std::size_t i) const {
+    return static_cast<value_type>(
+        reference(const_cast<uint32x2_t&>(m_value), int(i)));
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd_mask
+  operator||(simd_mask const& other) const {
+    return simd_mask(vorr_u32(m_value, other.m_value));
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd_mask
+  operator&&(simd_mask const& other) const {
+    return simd_mask(vand_u32(m_value, other.m_value));
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION simd_mask operator!() const {
+    auto const true_value = static_cast<uint32x2_t>(simd_mask(true));
+    return simd_mask(veor_u32(m_value, true_value));
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION bool operator==(
+      simd_mask const& other) const {
+    uint32x2_t const elementwise_equality = vceq_u32(m_value, other.m_value);
+    uint64x1_t const overall_equality_neon =
+        vreinterpret_u64_u32(elementwise_equality);
+    uint64_t const overall_equality = vget_lane_u64(overall_equality_neon, 0);
+    return overall_equality == 0xFFFFFFFFFFFFFFFFULL;
+  }
+  KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION bool operator!=(
+      simd_mask const& other) const {
+    return !operator==(other);
+  }
+};
+
+template <>
+class simd_mask<std::int64_t, simd_abi::neon_fixed_size<2>> {
+  uint64x2_t m_value;
```

Comment:
I cannot see the difference between the code in this `class` and the code in
`class simd_mask<double, simd_abi::neon_fixed_size<2>>`.  What is it?

Reply:
It may be exactly the same! they are both represented as `uint64x2_t` masks. We could make a common base class and try to derive, although we may have to use the Curiously Recurring Template Pattern to ensure the result types of operators are of the derived type

Reply:
Do we expect the two classes to always be the same? If we do, I think it would be useful to avoid duplicating code. Otherwise, we will need to fix a single bug multiple times.

Reply:
Okay! I've just pushed a couple of commits to de-duplicate this using CRTP base classes

Reply:
thanks